File: Algorithms/297_Serialize_and_Deserialize_Binary_Tree/Codec.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    private Iterator<String> it;
    private List<String> trace;

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        trace = new ArrayList<>();
        postOrder(root);
        return String.join(" ", trace);
    }

    // Decodes your encoded data to tree. 
    public TreeNode deserialize(String data) {
        TreeNode root = null;
        if (data != null && data.length() > 0) {
            String[] nodes = data.split(" ", -1);
            root = new TreeNode(Integer.parseInt(nodes[0].split("_", -1)[1]));
            it = Arrays.stream(nodes).iterator();
            build(root);
        }
        return root;
    }

    private void build(TreeNode root) {
        if (root != null) {
            String[] n = it.next().split("_", -1);
            root.left = n[0].length() > 0 ? new TreeNode(Integer.parseInt(n[0])) : null;
            root.right = n[2].length() > 0 ? new TreeNode(Integer.parseInt(n[2])) : null;
            build(root.left);
            build(root.right);
        }
    }

    private void postOrder(TreeNode root) {
        if (root != null) {
            trace.add(getVal(root.left) + "_" + root.val + "_" + getVal(root.right));
            postOrder(root.left);
            postOrder(root.right);
        }
    }

    private static String getVal(TreeNode node) {
        return "" + (node == null ? "" : node.val);
    }
}

File: Algorithms/449_Serialize_and_Deserialize_BST/Codec.scala
import scala.collection.mutable

/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 * var value: Int = _value
 * var left: TreeNode = null
 * var right: TreeNode = null
 * }
 */

class Codec {
  // Encodes a list of strings to a single string.
  def serialize(root: TreeNode): String = {
    val serialized = mutable.ListBuffer[String]()

    def preorder(node: TreeNode): Unit = {
      if (node != null) {
        serialized += s"${if (node.left == null) ""  else node.left.value}:${node.value}:${if (node.right == null) ""  else node.right.value}"
        preorder(node.left)
        preorder(node.right)
      }
    }

    preorder(root)
    serialized.mkString("_")
  }

  // Decodes a single string to a list of strings.
  def deserialize(data: String): TreeNode = if (data.isEmpty) null else {
    val q = data.split("_")
      .map(_.split(":", -1))
      .foldLeft(mutable.Queue[Array[String]]())(_.enqueue(_))
    val root = new TreeNode(q.head(1).toInt)

    def build(node: TreeNode): Unit = {
      if (node != null) {
        val next = q.dequeue()
        if (next(0).nonEmpty) node.left = new TreeNode(next(0).toInt)
        if (next(2).nonEmpty) node.right = new TreeNode(next(2).toInt)
        build(node.left)
        build(node.right)
      }
    }

    build(root)
    root
  }
}